effects into eff-start

cliu369 opened the pull request.

cliu369 committed:
Basic Binary Effect Tracking

cliu369 committed:
Added effect checking for applications

cliu369 committed:
Final version of simple binary effect system.

cliu369 committed:
Initial kill effect setup

cliu369 committed:
Basic well-formedness checks + checking DefDefs

cliu369 committed:
Small fix in DefDef

cliu369 committed:
Added preliminary kill effect to separation checker.

cliu369 committed:
Messing around with type level lists.

cliu369 committed:
Refactoring

Changed checking return value to seeing if return type captures any
killed value. Used markedFree in checking killed idents. Put all helper
methods into CheckEffects.

cliu369 committed:
Basic type inference infrastructure

cliu369 committed:
Basic checking done.

Basic checking in new type inference infrastructure finished.

cliu369 committed:
Basic inference + subtyping done

cliu369 committed:
Session types example

cliu369 committed:
Various fixes(fresh, subtyping, valdef)

1. Methods now check observable separation in separation checking - we
   set method's captured set to its captured variables in separation checker
2. Subtyping fix for killing function <: AppliedTypes
3. ValDef inference

cliu369 committed:
Small fixes

1. What is dead may not return.
2. Dealing with applying kill function to values
3. Well-formedness checks - requiring any killed variable to be a
   non-root capability.

cliu369 committed:
Mostly working recursive session types

I say mostly because of tuple problem with capture checking.

cliu369 committed:
More fixes

1. Prevent killing free variables (or at least of varialbes with owner
   above current owner). This is necessary for soundness until we allow
   function self-references.
2. Flow-sensitivity

cliu369 committed:
Preparation for use effect checking

1. Refactoring - moved various helper functions into new file
   EffectOps.scala and renamed only kill effect checker into KillChecker
2. Add onlyEffCheckKill function which serves as flag for whether only
   check kill or check use
3. Add infrastructure for checking use and kill.

cliu369 committed:
Basic use checking finished.

1. Checking that any use of killed variable is disallowed.
2. Checking that explicitly typed functions account for all killed/used
   parameters.

Tantalus13A98B5F commented on compiler/src/dotty/tools/dotc/core/TypeComparer.scala:
```diff
@@ -698,6 +699,49 @@ class TypeComparer(@constructorOnly initctx: Context) extends ConstraintHandling
+
+                  killSet1.subsetOf(killSet2) &&
+                    matchingMethodParams(info1, info2) &&
+                    isSubEff(info1.resultType.dropTopLevelKill, info2.resultType.subst(info2, info1).dropTopLevelKill)
```
Keeping note of our prior discussion:
1. kill sets are compared using `subsetOf`, instead of some qualifier relation.
2. I'm under the impression that we can only kill the arguments, not free variables, for now.
3. Argument comparison is done in `machingMethodParams`.

Tantalus13A98B5F commented on compiler/src/dotty/tools/dotc/core/TypeComparer.scala:
```diff
@@ -698,6 +699,49 @@ class TypeComparer(@constructorOnly initctx: Context) extends ConstraintHandling
+              // for poly should be like PolyType(args, MethodType(...))
+              case (info1: PolyType, info2: PolyType) =>
+                info1.paramNames.hasSameLengthAs(info2.paramNames)
+                && isSubEff(info1.resultType, info2.resultType.subst(info2, info1))
```
Are we comparing type bounds here?

cliu369 replied:
I don't think so - I guess that would be needed for comparing something like 
```def g[A <: (f: File => Unit @kill(f))](y: A) = { ... }```
This can be addition in future (I don't think this is most useful feature)?

Tantalus13A98B5F commented on compiler/src/dotty/tools/dotc/core/TypeComparer.scala:
```diff
@@ -1454,6 +1498,20 @@ class TypeComparer(@constructorOnly initctx: Context) extends ConstraintHandling
+          if isEffCheckingOrSetup && onlyEffCheckKill then
+              // if expected type is not dependent function type e.g. File^ => Unit
+              // and actual type is dependent function type with kill type e.g. (f: File^) => Unit @kill(f)
+              // then it should fail - in this case expected type is just an AppliedType
```
To further document the motivation here:

> This is because in Scala dependent function types are automatically subtypes of non-dependent function types (which are represented as AppliedTypes).
So by default in Scala something like
`(f: File^ => Unit @kill(f)) <: File^ => Unit`
is true. So this is to check for that case.

Tantalus13A98B5F commented on compiler/src/dotty/tools/dotc/cc/CheckCaptures.scala:
```diff
@@ -272,7 +275,15 @@ class CheckCaptures extends Recheck, SymTransformer:
 
-  def newRechecker()(using Context) = CaptureChecker(ctx)
+  // comment this out for original capture checker
+  var checker: CaptureChecker | Null = null
```
I had a question why you need to make this a global variable. Seems that you need that to initialize CheckEffects phase.

cliu369 replied:
Not good code but at the time I thought I needed the actual `CaptureChecker` object to get the all the information I needed in my phase, and so I made it global so that I can access the object outside of the capture checking phase.

Tantalus13A98B5F commented on compiler/src/dotty/tools/dotc/cc/SepCheck.scala:
```diff
@@ -773,13 +778,86 @@ class SepCheck(checker: CheckCaptures.CheckerAPI) extends tpd.TreeTraverser:
+  /**
+   * Used in traverseApply
+   */
+  private def checkKillApp(app: Apply, args: List[Tree])(using Context): Unit =
```
Probably orphan? I don't see where this method and below are used.

cliu369 replied:
This is orphan - I had originally implemented some checking directly in the separation checker and I kept it there as reference for my later code in separate phase.

Tantalus13A98B5F commented on compiler/src/dotty/tools/dotc/eff/FXSetup.scala:
```diff
@@ -0,0 +1,126 @@
+  private def updateInfo(sym: Symbol, info: Type)(using Context) =
+    sym.updateInfo(thisPhase, info, sym.flags)
+
+  class KillSetupTransformer(checker: CheckEffects.FXCheckerAPI) extends TreeMapWithPreciseStatContexts(cpy = cpyBetweenPhases):
```
I'm trying to summarize the job of this transformer -- "dropping all kills" on result types if they are inferred? What is the motivation here?

cliu369 replied:
Dropping all the kill annotations is not strictly necessary but the main point is that I want to perform type inference on inferred types, and so on these types I might as well just drop all the kill annotations since they will be inferred anyways and I don't want duplicates to happen.

Tantalus13A98B5F commented on compiler/src/dotty/tools/dotc/eff/EffectOps.scala:
```diff
@@ -0,0 +1,203 @@
+  extension (tree: Tree)
+    def killedElems(using Context): List[Tree] = tree match
+      case Apply(_, Typed(SeqLiteral(elems, _), _) :: Nil) =>
+        elems
```
~~This means all actual arguments involved in application are considered killed? I guess I need to look at the usages of this method.~~

I guess this specializes on the `Apply` node inside annotations. Other comments apply on `KillType`.

Tantalus13A98B5F commented on compiler/src/dotty/tools/dotc/eff/EffectOps.scala:
```diff
@@ -0,0 +1,203 @@
+
+  def apply(tp: Type, refs: List[Tree])(using Context): Type =
+    val annotTree =
+      New(getKillAnnot.typeRef,
```
In the construction form, we are creating a `New` node, while in the destruction form (`unapply` and then inside `killedElems`) we are looking for `Apply` nodes. So I guess there will be:

```scala
KillType(..., ...) match {
  case KillType(..., ...) => // mismatch!
```

Anything I'm missing?

cliu369 replied:
There is no mismatch actually - to be honest I'm not sure why there isn't (I haven't looked at it) because this is copied from how the capturing type gets its capture set from its annotation. 

My guess w/o looking at the code is that is because `New` is not the constructor for a `New` tree, but rather a wrapper method for constructing it which ends up making an `Apply` somehow.

Tantalus13A98B5F replied:
There we go, so it's an applied constructor.
https://github.com/TiarkRompf/scala3/blob/4d74cc63934f0b4d2cfdf012594df65d716e4b76/compiler/src/dotty/tools/dotc/ast/tpd.scala#L548-L555

Tantalus13A98B5F commented on compiler/src/dotty/tools/dotc/eff/EffectOps.scala:
```diff
@@ -0,0 +1,203 @@
+
+  /**
+   * Well-formedness conditions:1
+   * 1. Both sets cannot be empty (one set can be empty)
```
English(?): Sets cannot be both empty

Tantalus13A98B5F commented on compiler/src/dotty/tools/dotc/eff/EffectOps.scala:
```diff
@@ -0,0 +1,203 @@
+  end checkWellformed
+end KillOps
+
+object EffectType:
```
Is this the ongoing use effect extension? Not sure if it is currently utilized.

cliu369 replied:
Yes, although I'm probably going to work first on free variables in kill + function self ref first now (my logic is better to get kill correct first before doing use + kill since use + kill seems rather similar).

Tantalus13A98B5F commented on compiler/src/dotty/tools/dotc/eff/EffectOps.scala:
```diff
@@ -0,0 +1,203 @@
+    // This is the same as KillOps.killedElems
+    def killedElems: List[Tree] = tree match
+      case Apply(_, Typed(SeqLiteral(elems, _), _) :: Nil) => elems
+      case _ => Nil
```
Such `...Elems` methods seems generic, depending only on the annotation structure, and they themselves do not check for whether it is kill or effect. Perhaps could merge them and leave a note referring to `...Type.apply`.

cliu369 replied:
good idea!

Thanks for reviewing!

cliu369 committed:
Removed bineff

Tantalus13A98B5F commented on compiler/src/dotty/tools/dotc/eff/CheckEffects.scala:
```diff
@@ -0,0 +1,447 @@
+      case annot => annot
+
+  extension (sym: Symbol)
+    def isKill(using Context): Boolean =
```
Maybe you want to use `getKillAnnot` instead.

Tantalus13A98B5F commented on compiler/src/dotty/tools/dotc/eff/CheckEffects.scala:
```diff
@@ -0,0 +1,447 @@
+      atPhase(checkCapturesPhase)(getFP)
+
+    private def getFP(using Context): Refs = // todo make atPhase(CCPhase)
+      def recur(elems: Refs, newElems: List[CaptureRef]): Refs = newElems match
```
Transitive closure of captures?

cliu369 replied:
Should be transitive closure excluding top capability. Not 100% if its correct but seems to work so far.

Tantalus13A98B5F commented on compiler/src/dotty/tools/dotc/eff/CheckEffects.scala:
```diff
@@ -0,0 +1,447 @@
+     * 4. return new killed
+     *
+     * This is really inefficient - find a better solution.
+     * Probably the best way is to use something similar to ConsumedSet?
```
Sounds like you may want to use persistent sets or some kind of stack.

cliu369 replied:
I think the set needs to be mutable so I don't know if there is some nice persistent set/stack. Currently I'm thinking that I should just use the `mutable.HashSet` instead of the `util.HashSet` but fundamentally the data structure doesn't change. 

For their `@consume` effect they actually make a custom data structure `ConsumedSet` which uses an `Array` as the underlying data structure. I may follow them for this since I'm unsure about the performance characteristics of the Scala collections.

Tantalus13A98B5F replied:
What I was thinking about is,

```scala
var s = immutable.HashSet()
s = s + elem1
s = s + elem2
```

You can snapshot `s` above with rather low penalty I believe, but overall each update is going to be slower.

cliu369 replied:
That's a nice idea - I guess each individual update would copy the entire set which if large will also be costly.

Tantalus13A98B5F replied:
Not necessarily the entire set -- a good choice of the data structure can lower the cost to O(sqrt(n)) or O(log(n)).

e.g. https://usaco.guide/adv/persistent?lang=cpp

cliu369 replied:
I see. So depending on the underlying data structure of the immutable set, update can be done relatively quickly, but I imagine the mutable HashSet still has to be faster. 

So I guess it's a trade-off then - either faster `segment` with immutable or faster `update` with mutable.

Tantalus13A98B5F commented on compiler/src/dotty/tools/dotc/eff/CheckEffects.scala:
```diff
@@ -0,0 +1,447 @@
+          refs.flatMap(_.toCaptureRefs)
+        case _ => Nil
+
+      for param <- params do
```
let's say:

```scala
val foo = new Ref(42);
def bar(b: Ref[Int]^{foo}): Unit =
  kill(foo); ()
```

Should `b` be marked as killed on `bar`?

----

Ok down below "killing non-local is prohibited".

cliu369 replied:
That's another good point. My understanding is that `b` should be - since if its it not and `bar` escapes scope of `foo`, then `bar` will kill itself, but it may not necessarily kill `b` -
`Ref[Int]^{foo} => Unit @kill(foo)`  becomes `f: (Ref[Int]^{cap} => Unit @kill(f))`.

To check this I think we need to check if the parameter captures killed variables in the body in `checkExplicitDefDef`.

Tantalus13A98B5F commented on compiler/src/dotty/tools/dotc/eff/CheckEffects.scala:
```diff
@@ -0,0 +1,447 @@
+            ref.tpe match
+              case tp: CaptureRef if tp.isTrackableRef && !tp.isRootCapability => true
+              case _ => false
+          val deadRefs = SimpleIdentitySet(validRefs.flatMap(_.toCaptureRefs)*).footprint
```
So `killed` is implicitly transitively closed?

cliu369 replied:
Yes - I don't know if its more efficient with or without transitive closure but this one is maybe more robust so I just went with it for now.

Tantalus13A98B5F commented on compiler/src/dotty/tools/dotc/eff/CheckEffects.scala:
```diff
@@ -0,0 +1,447 @@
+      appType.dropTopLevelKill
+    end recheckApply
+
+    override def recheckClosureBlock(mdef: DefDef, expr: Closure, pt: Type)(using Context): Type =
```
What's the aim for having a `ClosureBlock` case? If it is for lambdas, why is there a `.symbol`?

cliu369 replied:
By default, every lambda in scala is lowered to 

```
def $anonfun() = { ... }
$anonfun
```
with a tree representation of `Block(DefDef(...), Closure)`. So this is to check for that case. 

The `DefDef` has a symbol so we use that here. I believe superclass `recheckClosureBlock` method won't perform type inference properly so I needed to make some changes for type inference to work (see my last comment for how inference is performed).

Tantalus13A98B5F commented on compiler/src/dotty/tools/dotc/eff/CheckEffects.scala:
```diff
@@ -0,0 +1,447 @@
+        newTpe
+    end recheckClosureBlock
+
+    override def recheckIf(tree: If, pt: Type)(using Context): Type =
```
Seems you can also use the `segment` function? Any reason we should not?

cliu369 replied:
I think I wrote `recheckIf` before I wrote segment. But also `recheckIf` only has two cases so maybe there is some optimization we can gain here? But yes `segment` should be totally usable here.

Tantalus13A98B5F commented on compiler/src/dotty/tools/dotc/eff/CheckEffects.scala:
```diff
@@ -0,0 +1,447 @@
+      val casesType = typesAndKill.map(_._1)
+      TypeComparer.lub(casesType)
+
+    override def recheckDef(tree: ValOrDefDef, sym: Symbol)(using Context): Type =
```
Why another case for `ValOrDefDef`, provided `ValDef` and `DefDef`? Any relations there?

cliu369 replied:
So `recheckDef` calls either `recheckValDef` or `recheckDefDef` depending on which one it is. 

In the template `Rechecker` (superclass of `KillChecker`), `recheckDef` is used in the main `recheck` match to check `ValDef` or `DefDef`. 

I think the main point of the method is to avoid code duplication between `recheckValDef` and `recheckDefDef` (see type inference comment at end).

cliu369 commented:
Type Inference Overview:
Every `DefDef` and `ValDef` which has had their type inferred needs to have type re-inferred at this phase to obtain the correct effect.

1. In `FXSetup` we set the info (type) of the denotation (phase indexed meaning of a symbol) of each inferred `DefDef` and `ValDef` to be a new LazyType with a completer that calls `recheckDef` from the actual checking phase. 
2. When we want to get the type of a `DefDef` or `ValDef` we ask for the info of its denotation (`sym.info`). This will check if it is a `LazyType`. If so, this means it has not yet been inferred and the completer will run, which will call `recheckDef` to check it again. 
3. The `sym.info` can also be completed by calling `sym.ensureCompleted()`, and can also be obtained by calling `tree.tpe.widen` on a tree associated with the symbol (although the latter one depends on the specific type). 
4. To ensure that we do not try to check the same tree twice, after we finish `recheckDef` of a tree, we add its symbol to a `completed` set, and if a symbol is already in there then it won't be checked again.

cliu369 committed:
Basic tests for kill based on automation.

1. Run via sbt using `testCompilation typestate`

cliu369 committed:
-Yrecheck-test fix for detecting kill annotation in tests

Other small things:
1. Begin function self-ref/killing free variables
2. Using mutable.HashSet instead of util.HashSet

cliu369 committed:
Prevent running of TestRecheck Phase

cliu369 committed:
Steps to killing free variables

1. Prevent free variables from being killed inside while loop
2. Added inference and checking to DefDef for free variables
3. Supported => T (exprType)

cliu369 committed:
Basic avoidance

cliu369 committed:
Deleted orphan separation checking code
